developer_guides/uuid/index.rst

```diff
@@ -0,0 +1,134 @@
+.. _uuid:
+
+UUID Usage in SOF
+#################
+
+Why UUID is needed
+******************
+
+When developing a new audio signal processing component, we
+traditionally implement the component driver with a new unique component
+type declared together, the painpoints with this method including ABI
+alignment among topology/driver/firmware for each new added components,
+and there could be collision if versions are not aligned.
+
+UUIDs (Universally Unique Identifiers) provide a more scalable and
+collisionn-free way of component identification. UUIDs are currently
+used as the standard interface by all users of the SOF firmware,
+including the tracing subsystem, the topology .m4 files, and the Linux
+topology driver.
+
+Allocating a new UUID for the new added component is mandatory, and
+UUIDs are expected to replace the component type in the near future. For
+the whole SOF stack, the usage of UUID should follow rules as below:
+
+UUID allocation
+***************
+The UUID allocation of a specific component should use the version 4 in
+https://en.wikipedia.org/wiki/Universally_unique_identifier, and the
+value should be specified in the firmware component dirver, e.g. for src
+component, in sof/src/audio/src.c:
+
+.. code-block:: none
+
+   /* c1c5326d-8390-46b4-aa47-95c3beca6550 */
+   DECLARE_SOF_UUID("src", src_uuid, 0xc1c5326d, 0x8390, 0x46b4,
+                    0xaa, 0x47, 0x95, 0xc3, 0xbe, 0xca, 0x65, 0x50);
+
+UUID in topology
+****************
+We should provide UUID string in topology .m4 file in the exactly same
+format as commented in the firmware, e.g for host it is
+c1c5326d-8390-46b4-aa47-95c3beca6550, so in the topology .m4 file
+tools/topology/m4/src.m4:
+
+.. code-block:: none
+
+   dnl Defines the macro for SRC widget
+   define(SRC_UUID_ARRAY, UUID_ARRAY("c1c5326d-8390-46b4-aa47-95c3beca6550"))
```

Comment:
this makes no sense to me, why not use the same format as line 35 and 36 so that we can copy/paste between firmware and topology?

Reply:
line 35 and 36 is C macro to define struct sof_uuid_entry, but ASoC UUID token is u8 array[16], what did you suggest to copy/paste? the "0xc1c5326d, 0x8390, 0x46b4, 0xaa, 0x47, 0x95, 0xc3, 0xbe, 0xca, 0x65, 0x50)"? The UUID standard (see https://en.wikipedia.org/wiki/Universally_unique_identifier) string "c1c5326d-8390-46b4-aa47-95c3beca6550" (copy/paste from the line 34) looks more compact and reasonable, no?

Reply:
I don't care about the format, I just want the firmware files and topoligy files to use the same format to avoid having to edit bytes.

Reply:
> I don't care about the format, I just want the firmware files and topology files to use the same format to avoid having to edit bytes.

That was what @lgirdwood commented, we already support copying from the firmware files and pasting to the topology .m4 files directly, no need to edit bytes.

Reply:
Line 35-36 and line 48 show a different format so there's no copy-paste possible between the firmware code and the topology file.

Reply:
> Line 35-36 and line 48 show a different format so there's no copy-paste possible between the firmware code and the topology file. the fact that the internal representation is identical with a 16-byte array is irrelevant, what we need to have is the same macro with the same split between bytes.

If you can treat Line *34*(the comment line) and line 48, they are in same format. As you know, here they are C and m4, it is not easy to implement the C structure definition in m4 with the same macros. @plbossart are you suggesting something like this?
```
UUID_ARRAY("0xc1c5326d, 0x8390, 0x46b4, 0xaa, 0x47, 0x95, 0xc3, 0xbe, 0xca, 0x65, 0x50")
```

Reply:
This is really simple. We either use

format a) c1c5326d-8390-46b4-aa47-95c3beca6550
format b) 0xc1c5326d, 0x8390, 0x46b4,
                    0xaa, 0x47, 0x95, 0xc3, 0xbe, 0xca, 0x65, 0x50

Pick one for both firmware and M4 macro, don't ask people to format themselves.

Reply:
@keyonjie I think we agreed 
1) to use the same definition in FW and topology (to support simple copy and paste).
2) to use the same binary ABI representation between topology and FW.
Please setup a call with me if unsure.

Reply:
the PR is updated with this part changed, mark it 'resolved'.